Remove legacy user fallback

kobelb opened the pull request:
**ATTN teams accidentally tagged for review as code owners, I messed up a rebase. Please ignore.**


When we introduced RBAC and Kibana Application privileges in 6.4, we added support to use the "legacy fallback" when users didn't have any Kibana Application privileges. In this scenario, when the user logs into Kibana, we're logging a deprecation warning but allowing them to continue relying upon direct index privileges.

We're removing this capability in 7.0, and users will now be required to have roles which grant them Kibana Application privileges.

This unblocks features like https://github.com/elastic/kibana/issues/18270

"Relate Note: Removed legacy fallback which allows users to rely upon direct index access to the Kibana index in Elasticsearch. Users should switch to using Kibana privileges"

kobelb committed:
Remove mode.initialize and change useRbacForRequest to useRbac

kobelb committed:
Updating saved object api tests

kobelb committed:
Fixing spaces api integration tests

elasticmachine commented:
Pinging @elastic/kibana-security

kobelb committed:
Removing unused "expect legacy forbidden" declarations and imports

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7141/)

kobelb committed:
Updating docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7158/)

kobelb commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7173/)

kobelb commented:
Last test run was against https://github.com/kobelb/elasticsearch/tree/remove-legacy-fallback which incorporated some recent ES changes which Kibana doesn't support yet.

kobelb commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7186/)

legrego commented on docs/migration/migrate_7_0.asciidoc:
```diff
@@ -9,6 +9,13 @@ your application to Kibana 7.0.
+*Details:* With the introduction of Kibana RBAC in 6.4, user's no longer require privileges to the Kibana index in Elasticsearch. Instead, users 
+should be granted <<kibana-privileges>>. Prior to 7.0, when a user that relies upon direct index privileges logs into Kibana, a deprecation warning is logged. If you are using the `kibana_user` role to grant access to Kibana, or a custom role using <<kibana-privileges>>, no changes are required.
+
+*Impact:* You must change any roles which grant access to Kibana using index privileges to instead use <<kibana-privileges>>.
```
It's definitely implied, but what do you think about calling out reporting jobs here? Jobs with auth setup will need to ensure that those accounts are updated as well. Otherwise, they'll start failing with a non-descriptive "forbidden" error.

kobelb replied:
> Jobs with auth setup will need to ensure that those accounts are updated as well. Otherwise, they'll start failing with a non-descriptive "forbidden" error.

Are you referring to pending jobs that are using the legacy roles at the time of upgrade? Or is there some other scenario which I'm not thinking of?

legrego replied:
I don't care so much about the pending jobs, as they should be negligible. I was more concerned about configured watcher actions. Users can configure Watcher alerts to [include a Kibana report as an email attachment](https://www.elastic.co/guide/en/elastic-stack-overview/6.5/actions-email.html).

legrego reviewed (commented):
Tested locally, this is working great! Just a couple of small comments.

legrego committed:
Update docs/migration/migrate_7_0.asciidoc

Co-Authored-By: kobelb <brandon.kobel@gmail.com>

legrego committed:
Update docs/migration/migrate_7_0.asciidoc

Co-Authored-By: kobelb <brandon.kobel@gmail.com>

kobelb commented on x-pack/plugins/security/index.js:
```diff
@@ -43,17 +43,18 @@ export const security = (kibana) => new kibana.Plugin({
     }).default();
   },
 
+  deprecations: function ({ unused }) {
```
Yeah, it's not supposed to crash, lemme look into this.

kobelb committed:
Updating comment that mentions the scenario when we aren't using RBAC

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7247/)

kobelb committed:
Adding back the authorization section of the config

When a config setting is marked as unused using the deprecations, it's
still required to show up in the config declarations so an error isn't
thrown on startup.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7254/)

kobelb committed:
Adding note about watcher jobs

legrego commented on docs/migration/migrate_7_0.asciidoc:
```diff
@@ -9,6 +9,13 @@ your application to Kibana 7.0.
 See also <<release-highlights>> and <<release-notes>>.
 
+[float]
+=== Removed support for user's relying on direct index privileges to the Kibana index in Elasticsearch
```
Sorry, missed one:
```suggestion
=== Removed support for users relying on direct index privileges to the Kibana index in Elasticsearch
```

kobelb replied:
Thanks... I don't know what happened to my grammar/english skills.

legrego committed:
Update docs/migration/migrate_7_0.asciidoc

Co-Authored-By: kobelb <brandon.kobel@gmail.com>

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7282/)

legrego commented:
retest

kobelb commented:
ATTN teams accidentally tagged for review as code owners, I messed up a rebase. Please ignore.

kobelb committed:
Merge remote-tracking branch 'upstream/master' into remove-legacy-fallback

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7575/)

kobelb committed:
Merge remote-tracking branch 'upstream/master' into remove-legacy-fallback

kobelb commented:
Closing in favor of https://github.com/elastic/kibana/pull/29107 to get CI green.
